ledger/src/blockstore_processor.rs

```diff
@@ -91,387 +90,183 @@ impl BlockCostCapacityMeter {
-                )
-            );
-        }
-    }
-    first_err
+#[derive(Default)]
+struct ExecuteBatchesInternalMetrics {
+    execute_batches_us: u64,
 }
 
-fn aggregate_total_execution_units(execute_timings: &ExecuteTimings) -> u64 {
-    let mut execute_cost_units: u64 = 0;
-    for (program_id, timing) in &execute_timings.details.per_program_timings {
-        if timing.count < 1 {
-            continue;
-        }
-        execute_cost_units =
-            execute_cost_units.saturating_add(timing.accumulated_units / timing.count as u64);
-        trace!("aggregated execution cost of {:?} {:?}", program_id, timing);
-    }
-    execute_cost_units
+#[derive(Clone)]
+enum ExecutionState {
+    Blocked,
+    Processing,
+    Processed,
 }
 
-fn execute_batch(
-    batch: &TransactionBatchWithIndexes,
+#[allow(clippy::too_many_arguments)]
+fn execute_batches_internal(
     bank: &Arc<Bank>,
+    transactions_indices_to_schedule: &[(SanitizedTransaction, usize)],
+    entry_callback: Option<&ProcessCallback>,
     transaction_status_sender: Option<&TransactionStatusSender>,
     replay_vote_sender: Option<&ReplayVoteSender>,
-    timings: &mut ExecuteTimings,
+    confirmation_timing: &mut ConfirmationTiming,
     cost_capacity_meter: Arc<RwLock<BlockCostCapacityMeter>>,
-    tx_cost: u64,
+    tx_costs: &[u64],
     log_messages_bytes_limit: Option<usize>,
-) -> Result<()> {
-    let TransactionBatchWithIndexes {
-        batch,
-        transaction_indexes,
-    } = batch;
-    let record_token_balances = transaction_status_sender.is_some();
-
-    let mut mint_decimals: HashMap<Pubkey, u8> = HashMap::new();
+    replayer_handle: &ReplayerHandle,
+) -> Result<ExecuteBatchesInternalMetrics> {
+    assert!(!transactions_indices_to_schedule.is_empty());
 
-    let pre_token_balances = if record_token_balances {
-        collect_token_balances(bank, batch, &mut mint_decimals)
-    } else {
-        vec![]
-    };
+    let mut num_left_to_process = transactions_indices_to_schedule.len();
+    let mut processing_state =
+        vec![ExecutionState::Blocked; transactions_indices_to_schedule.len()];
+    let mut transaction_results = vec![Ok(()); transactions_indices_to_schedule.len()];
 
-    let pre_process_units: u64 = aggregate_total_execution_units(timings);
+    let mut execute_batches_elapsed = Measure::start("execute_batches_elapsed");
 
-    let (tx_results, balances) = batch.bank().load_execute_and_commit_transactions(
-        batch,
-        MAX_PROCESSING_AGE,
-        transaction_status_sender.is_some(),
-        transaction_status_sender.is_some(),
-        transaction_status_sender.is_some(),
-        transaction_status_sender.is_some(),
-        timings,
-        log_messages_bytes_limit,
-    );
+    // build a resource-based dependency graph
+    let tx_account_locks_results: Vec<Result<_>> = transactions_indices_to_schedule
+        .iter()
+        .map(|(tx, _)| tx.get_account_locks(MAX_TX_ACCOUNT_LOCKS))
```

Comment:
should we use `bank.get_transaction_account_lock_limit()` here?

Reply:
yes